src/main/java/net/minecraftforge/common/network/NetworkChannelHelper.java

```diff
@@ -0,0 +1,129 @@
+ */
+
+package net.minecraftforge.common.network;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelOutboundInvoker;
+import io.netty.channel.ChannelPromise;
+import net.minecraft.network.INetHandler;
+import net.minecraft.network.play.INetHandlerPlayClient;
+import net.minecraft.network.play.INetHandlerPlayServer;
+import net.minecraftforge.common.ForgeModContainer;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+import net.minecraftforge.fml.common.gameevent.TickEvent;
+import net.minecraftforge.fml.common.gameevent.TickEvent.Phase;
+
+public class NetworkChannelHelper
+{
+    private static Set<ChannelOutboundInvoker> markedChannels;
+    
+    static
+    {
+        markedChannels = new HashSet<ChannelOutboundInvoker>();
+        Thread t = new Thread(NetworkChannelHelper::run, "NetworkChannelFlush");
+        t.setDaemon(true);
+        t.start();
+    }
+    
+    
+    
+    public static ChannelFuture writeAndFlush(ChannelOutboundInvoker channel, Object msg, INetHandler packetListener)
+    {
+        if(packetListener==null)
+        {
+            return channel.writeAndFlush(msg);
+        }
+        else if(packetListener instanceof INetHandlerPlayClient || packetListener instanceof INetHandlerPlayServer)
+        {
+            return writeAndFlush(channel, msg);
+        }
+        else
+        {
+            return channel.writeAndFlush(msg);
+        }
+    }
+    
+    public static ChannelFuture writeAndFlush(ChannelOutboundInvoker channel, Object msg)
+    {
+        if(isInstantFlushEnabled())
+        {
+            return channel.writeAndFlush(msg);
+        }
+        else
+        {
+            markForFlush(channel);
+            return channel.write(msg);
+        }
+    }
+    
+    public static ChannelFuture writeAndFlush(ChannelOutboundInvoker channel, Object msg, ChannelPromise promise)
+    {
+        if(isInstantFlushEnabled())
+        {
+            return channel.writeAndFlush(msg, promise);
+        }
+        else
+        {
+            markForFlush(channel);
+            return channel.write(msg, promise);
+        }
+    }
+    
+    public static void markForFlush(ChannelOutboundInvoker channel)
+    {
+        synchronized (markedChannels)
+        {
+            markedChannels.add(channel);
+        }
+    }
+    
+    public static void flushAllChannels()
+    {
+        synchronized (markedChannels)
+        {
+            markedChannels.forEach(ChannelOutboundInvoker::flush);
+            markedChannels.clear();
+        }
+    }
+    
+    private static boolean isInstantFlushEnabled()
+    {
+        return ForgeModContainer.instantNetworkFlushing;
+    }
+    
+    public static void run()
+    {
+        try
+        {
+            while(true)
+            {
+                flushAllChannels();
```

Comment:
Is it possible for this to happen from the main server thread in the actual tick loop, or would flushing be too expensive there? How does Vanilla achieve this?

Reply:
avnilla flush by itself, this is to make sure even when TPS is low packets get send, normal this can be decactivated and it will still work. (I tested this)

Reply:
If this thread is enabled I imagine vanilla flushing should be disabled. Otherwise performance improvements from this PR might be reduced due to flushing still happening twice a tick, once from vanilla and once here.